src/test/ui/traits/trait-param-without-lifetime-constraint.stderr

```diff
@@ -0,0 +1,21 @@
+error: `impl` item signature doesn't match `trait` item signature
```

Comment:
hmm, so I personally still find this fairly confusing :(

The current output is

```
error: `impl` item signature doesn't match `trait` item signature
  --> src/main.rs:14:5
   |
10 |     fn get_relation(&self) -> To;
   |     ----------------------------- expected fn(&Article) -> &ProofReader
...
14 |     fn get_relation(&self) -> &ProofReader {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ found fn(&Article) -> &ProofReader
   |
   = note: expected `fn(&Article) -> &ProofReader`
              found `fn(&Article) -> &ProofReader`
```

the point here is to warn folks that they may have wanted `-> &To` in the trait, though if they do that, they also need to change the impl to `HaveRelationship<ProofReader>`, right? (which the new error doesn't hint at)

Hmm, this is tricky.

Reply:
I think my ideal message might look something like this

```
error: `impl` item signature doesn't match `trait` item signature
  --> src/main.rs:14:5
   |
10 |     fn get_relation(&self) -> To;
   |                               -- maybe you want `&To` here instead?
...
13 |  impl HaveRelationship<&ProofReader> for Article { 
   |                        - let's call this lifetime `'1`
14 |     fn get_relation(&self) -> &ProofReader {
   |                     -         ^^^^^^^^^^^^ impl returns reference with lifetime `'2` but trait declares lifetime `'1`
   |                     |
   |                     let's call this lifetime `'2`
   |
   = note: expected `fn(&Article) -> &ProofReader`
              found `fn(&Article) -> &ProofReader`
```

(ignoring the fact that the "maybe you want" should ideally come as a separate chunk at the end)

Reply:
I'm looking at this again. Right now the output is:

![Screen Shot 2020-03-29 at 4 59 52 PM](https://user-images.githubusercontent.com/1606434/77864656-be715c00-71de-11ea-8ec9-c96ff9884ded.png)

I'm looking at what options I might have to approximate the suggested output, but it can quickly get tricky when more complex lifetime relationships are used.

Reply:
I'm finding doing the same kind of fancy pointing of lifetimes as we do in `librustc_mir::borrowck` a bit hard to do for this. It seems like I might have to write the logic to associate HIR lifetimes to `ty::Region` myself, but most likely I'm just missing something obvious somewhere.

Reply:
I'm not sure how easy it will be to repurpose that logic indeed. Let me see if I can give you any tips :)

Reply:
Hmm, OK, so I am reminded that there are still some bits of the code that use the lexical region solver. I wonder if the "right fix" is to be moving those bits of code to use the NLL solver, which might then enable us to use the NLL rewriting routines more easily. This is work that has been very slowly proceeding, @matthewjasper was carrying it along to some extent.